File: flutter-sdk/lib/src/ui/chat/view_models/audio/audio_state.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

import 'package:chat_flutter_sdk/src/domain/models/audio/audio_data.dart';
import 'package:chat_flutter_sdk/src/domain/models/chat_message/chat_message.dart'
    show ChatMessage;
import 'package:equatable/equatable.dart';

enum AudioStatus {
  errorStoppingAudio,
  errorPlayingAudio,
  errorStoppingRecording,
  errorRecordingAudio,
  recordingAudio,
  playingAudio,
  audioPaused,
  initial,
}

final class AudioState extends Equatable {
  final AudioData audioData;
  final int amplitudeIndex;
  final bool isUserRecordingAudio;
  final AudioStatus audioStatus;

  final ChatMessage? playingMessage;

  AudioState({
    AudioData? audioData,
    this.amplitudeIndex = 0,
    this.isUserRecordingAudio = false,
    this.playingMessage,
    this.audioStatus = AudioStatus.initial,
  }): audioData = audioData ?? AudioData();

  AudioState copyWith({
    AudioData? audioData,
    int? amplitudeIndex,
    bool? isUserRecordingAudio,
    ChatMessage? Function()? playingMessage,
    AudioStatus? audioStatus,
  }) {
    return AudioState(
      audioData: audioData ?? this.audioData,
      amplitudeIndex: amplitudeIndex ?? this.amplitudeIndex,
      isUserRecordingAudio: isUserRecordingAudio ?? this.isUserRecordingAudio,
      playingMessage: playingMessage != null
          ? playingMessage()
          : this.playingMessage,
      audioStatus: audioStatus ?? this.audioStatus,
    );
  }

  @override
  List<Object?> get props => [
    audioData,
    amplitudeIndex,
    isUserRecordingAudio,
    playingMessage,
    audioStatus,
  ];
}

File: android-sdk/sdk/src/main/kotlin/com/yalo/chat/sdk/ui/chat/AudioState.kt
// Copyright (c) Yalochat, Inc. All rights reserved.

package com.yalo.chat.sdk.ui.chat

import com.yalo.chat.sdk.domain.model.AudioData
import com.yalo.chat.sdk.domain.model.ChatMessage
import com.yalo.chat.sdk.ui.chat.AudioViewModel.Companion.AMPLITUDE_DATA_POINTS
import com.yalo.chat.sdk.ui.chat.AudioViewModel.Companion.DEFAULT_AMPLITUDE

sealed class AudioStatus {
    data object Initial : AudioStatus()
    data object RecordingAudio : AudioStatus()
    data object PlayingAudio : AudioStatus()
    data object AudioPaused : AudioStatus()
    data object ErrorRecordingAudio : AudioStatus()
    data object ErrorStoppingRecording : AudioStatus()
    data object ErrorPlayingAudio : AudioStatus()
    data object ErrorStoppingAudio : AudioStatus()
}

data class AudioState(
    val audioData: AudioData = AudioData(
        amplitudes = List(AMPLITUDE_DATA_POINTS) { DEFAULT_AMPLITUDE },
        amplitudesPreview = List(AMPLITUDE_DATA_POINTS) { DEFAULT_AMPLITUDE },
    ),
    // Index drives the sliding-waveform animation, decremented on each amplitude tick.
    val amplitudeIndex: Int = AMPLITUDE_DATA_POINTS - 1,
    val playingMessage: ChatMessage? = null,
    val audioStatus: AudioStatus = AudioStatus.Initial,
)

// Derived convenience — avoids a redundant boolean field that could disagree with audioStatus.
val AudioState.isRecording: Boolean get() = audioStatus is AudioStatus.RecordingAudio
